components/butterfly/src/election.rs

```diff
@@ -0,0 +1,257 @@
+    }
+
+    /// Insert a vote for the election.
+    pub fn insert_vote(&mut self, member_id: &str) {
+        if !self.get_votes().contains(&String::from(member_id)) {
+            self.mut_votes().push(String::from(member_id));
+        }
+    }
+
+    /// Steal all the votes from another election for ourselves.
+    pub fn steal_votes(&mut self, other: &mut Election) {
+        for x in other.mut_votes().iter() {
+            self.insert_vote(x);
+        }
+    }
+
+    /// Sets the status of the election to "running".
+    pub fn running(&mut self) {
+        self.set_status(Election_Status::Running);
+    }
+
+    /// Sets the status of the election to "finished"
+    pub fn finish(&mut self) {
+        self.set_status(Election_Status::Finished);
+    }
+
+    /// Sets the status of the election to "NoQuorum"
+    pub fn no_quorum(&mut self) {
+        self.set_status(Election_Status::NoQuorum);
+    }
+
+    /// Returns true if the election is finished.
+    pub fn is_finished(&self) -> bool {
+        self.get_status() == Election_Status::Finished
+    }
+}
+
+impl PartialEq for Election {
+    /// We ignore id in equality checking, because we only have one per service group
+    fn eq(&self, other: &Election) -> bool {
+        self.get_service_group() == other.get_service_group() &&
+        self.get_member_id() == other.get_member_id() &&
+        self.get_suitability() == other.get_suitability() &&
+        self.get_votes() == other.get_votes() && self.get_status() == other.get_status() &&
+        self.get_term() == other.get_term()
+    }
+}
+
+impl Rumor for Election {
+    /// Updates this election based on the contents of another election.
+    fn merge(&mut self, mut other: Election) -> bool {
+        if *self == other {
+            // If we are the same object, just return false
+            // println!("Equal: {:?} {:?}", self, other);
+            false
+        } else if other.get_term() >= self.get_term() &&
+                  other.get_status() == Election_Status::Finished {
+            // If the new rumors term is bigger or equal to ours, and it has a leader, we take it as the
+            // leader and move on.
+            *self = other;
+            true
+        } else if other.get_term() == self.get_term() &&
+                  self.get_status() == Election_Status::Finished {
+            // If the terms are equal, and we are finished, then we drop the other side on the
+            // floor
+            false
+        } else if self.get_term() > other.get_term() {
+            // If the rumor we got has a term that's lower than ours, keep sharing our rumor no
+            // matter what term they are on.
+            true
+        } else if self.get_suitability() > other.get_suitability() {
+            // If we are more suitable than the other side, we want to steal
+            // the other sides votes, and keep sharing.
+            // println!("Self suitable: {:?} {:?}", self, other);
```

Comment:
`debug()!` here as well? There's a few below in this module if you want to take a peek at those, too :)